### DISCUSSION TITLE
Log4j should stop logging when the disk is full.

### DISCUSSION THREAD

**Discussion by rupesh1306** [apache/logging-log4j2#2548] [Q&A]
Hi,
Greetings!
Can we stop Log4j to log any event when the disk is about to full?
Is there any configuration in Log4J which can do this?
Is there any way we can have the above mentioned behavior? 
is there a way if we can have per day log file size limit? for example per day limit is set 2GB then after 2GB it should not log the events in log file.

Thanks in Advance,
Rupesh

**Comment by vy** [comment]
> Can we stop Log4j to log any event when the disk is about to full?

Yes, I know of the following method:

1. You can either shutdown Log4j programmatically (using `LogManager.shutdown()`) or _"reconfigure"_ it to discard all logging. The reconfiguration can be triggered either via JMX or programmatically using `Configurator.reconfigure()` methods.
1. Either your system administrator can invoke these methods through a (REST?) API or you can write a background task to monitor the disk usage and trigger it programmatically.

> Is there any configuration in Log4J which can do this?

Not that I know of.

> is there a way if we can have per day log file size limit?

Not that I know of.

**Reply by rgoers** [reply]
That isn't quite accurate. Using the SizeBasedTriggeringPolicy you can cause the file to roll over based on it reaching a certain size. You can specify the number of files that are allowed.  So if you allow a file to grow to 30MB before rollover and then limit it to 10 files then the max will be 300MB. If used in conjunction with a time-based rollover policy it gets tricker as the maxFiles only applies to each time interval. In that case you need to use a delete action to limit the number of files.

**Comment by ppkarwasz** [comment]
@rupesh1306,

> is there a way if we can have per day log file size limit? for example per day limit is set 2GB then after 2GB it should not log the events in log file.

Recent log events are usually more valuable than old events, so you don't want to stop writing logs after 2 GiB are written, you probably want to delete the old ones. For this there is a standard solution: [rolling file appender](https://logging.apache.org/log4j/2.x/manual/appenders#rollingfileappender). You basically need a configuration like this:

```xml
<RollingFile name="ROLLING"
             filePattern="logs/app.%d{yyyy-MM-dd}.%i.log">
  <DirectRolloverStrategy maxFiles="4"/>
  <Policies>
    <TimeBasedTriggeringPolicy/>
    <SizeBasedTriggeringPolicy size="512 M"/>
  </Policies>
</RollingFile>
```

This way you'll get the most recent 2 GiB of logs per day.

**Remark**: I don't how big your system is, but if you are getting 2 GiB of logs per day, you are probably logging too much.
You could use multiple appenders to:

* Log the important messages (e.g. `WARN` and less specific) to one file that will be stored for a **long** period,
* Log debug messages (`DEBUG` and less specific) to another rolling file that will retain only one or two archives.

**Reply by rgoers** [reply]
As I noted above, with this configuration you really should also have a Delete action so that you don't keep files over x time intervals old. Otherwise it could still grow endlessly.
